### DISCUSSION TITLE
How to use serwist in my existing next.js/t3 app?

### DISCUSSION THREAD

**Discussion by FleetAdmiralJakob** [serwist/serwist#8] [Q&A]
I have already an T3 app and want to add serwist / a PWA. How can I do this with serwist?

**Comment by DuCanhGH** [comment]
@FleetAdmiralJakob You only need to add app/sw.ts:

```ts
import { defaultCache } from "@serwist/next/browser";
import type { PrecacheEntry } from "@serwist/precaching";
import { installSerwist } from "@serwist/sw";

declare const self: ServiceWorkerGlobalScope & {
  // Change this attribute's name to your `injectionPoint`.
  __SW_MANIFEST: (PrecacheEntry | string)[] | undefined;
};

installSerwist({
  precacheEntries: self.__SW_MANIFEST,
  skipWaiting: true,
  clientsClaim: true,
  navigationPreload: true,
  runtimeCaching: defaultCache,
});
```

and change your next.config.mjs:

```js
// @ts-check
import withSerwistInit from "@serwist/next";

const withSerwist = withSerwistInit({
  cacheOnFrontEndNav: true,
  swSrc: "app/sw.ts",
  swDest: "public/sw.js",
});

/** @type {import("next").NextConfig} */
const nextConfig = {
  reactStrictMode: true,
};

export default withSerwist(nextConfig);
```

**Reply by I-Atlas** [reply]
@DuCanhGH, Is serwist working with pages directory?

**Reply by DuCanhGH** [reply]
@I-Atlas probably. If you have issues using Serwist with the `pages/` directory, do open an issue :)

**Reply by I-Atlas** [reply]
But how to use it with pages?

**Reply by FleetAdmiralJakob** [reply]
@I-Atlas I think the same. The explanation was not app dir specific.

**Reply by DuCanhGH** [reply]
@FleetAdmiralJakob well, that `swSrc` is set to "app/sw.ts" is pretty App Router-specific if you ask me :) So you'd have to change that if you are only using Pages Router.

**Reply by I-Atlas** [reply]
@DuCanhGH, and what is meant by `injectionPoint` ? 😅

**Reply by DuCanhGH** [reply]
@I-Atlas it is a Serwist option, which dictates the string that `@serwist/webpack-plugin` should find and replace with the list of precache entries.

```js
// @ts-check
import withSerwistInit from "@serwist/next";

const withSerwist = withSerwistInit({
  cacheOnFrontEndNav: true,
  swSrc: "app/sw.ts",
  swDest: "public/sw.js",
  injectionPoint: "self.__LOREM",
});

/** @type {import("next").NextConfig} */
const nextConfig = {
  reactStrictMode: true,
};

export default withSerwist(nextConfig);
```

```ts
import { defaultCache } from "@serwist/next/browser";
import type { PrecacheEntry } from "@serwist/precaching";
import { installSerwist } from "@serwist/sw";

declare const self: ServiceWorkerGlobalScope & {
  // Change this attribute's name to your `injectionPoint`.
  __LOREM: (PrecacheEntry | string)[] | undefined;
};

installSerwist({
  precacheEntries: self.__LOREM,
  skipWaiting: true,
  clientsClaim: true,
  navigationPreload: true,
  runtimeCaching: defaultCache,
});
```

**Reply by FleetAdmiralJakob** [reply]
For any people after us and looking for a solution: You have to add `webworker` to your libs in the tsconfig and you have to add a link to your manifest in the layout or _app file

**Comment by ajayvignesh01** [comment]
> For any people after us and looking for a solution: You have to add `webworker` to your libs in the tsconfig and you have to add a link to your manifest in the layout or _app file

what is the purpose of adding the webworker to ts config?

**Reply by DuCanhGH** [reply]
@0x-Legend its purpose is to make WebWorker-specific types like `ServiceWorkerGlobalScope` accessible :)
